### PR TITLE
TIS-565/Fintraffic UI customisations

### PR DESCRIPTION
**Pull request by solita-sabinaf**
Customisation areas:

- Styles
- App logo
- Translations, e.g. overriding certain values or additional company-specific keys

### CODE REVIEW

File: src/config/config.ts [hunk 1]
```diff
@@ -9,7 +9,9 @@ import { Locale } from '../i18n';
  *  For multi-level features, only the top-level featureName should be
  *  toggled.
  */
-export interface SandboxFeatures {}
+export interface SandboxFeatures {
+  Fintraffic: boolean;
+}
```
Comment (solita-sabinaf) [inline src/config/config.ts:14]: The idea was to make a root Fintraffic folder, so that everything entirely company-specific stays there

File: src/ext/Fintraffic/assets/manifest.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    {
+      "src": "favicon.ico",
+      "sizes": "32x32"
+    }
+  ],
+  "start_url": "./index.html",
+  "display": "standalone",
+  "theme_color": "#000000",
+  "background_color": "#ffffff"
+}
```
Comment (solita-sabinaf) [inline src/ext/Fintraffic/assets/manifest.json:14]: This felt questionable, we could also override this at a build time. But maybe it's easier this way, just a link to plug in under <Helmet />

File: src/ext/Fintraffic/translations/en.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export const messages = {
+  appTitle: 'RAE',
+  appLongTitle: 'Route and schedule editor',
+};
```
Comment (solita-sabinaf) [inline src/ext/Fintraffic/translations/en.ts:4]: appLongTitle is only used in Fintraffic's logo feature

File: src/helpers/validation.ts [hunk 4]
```diff
@@ -245,13 +245,6 @@ export const validateServiceJourney = (
   const validDayTimes = (sj.dayTypes?.[0]?.daysOfWeek?.length ?? 0) > 0;
   const { isValid } = validateTimes(sj.passingTimes ?? [], intl);
   const validDayTypes = validateDayTypes(sj.dayTypes);
-  console.log('validateServiceJourney', {
-    sj,
-    validName: !isBlankName,
-    validDayTimes,
-    validPassingTimes: isValid,
-    validDayTypes,
-  });
```
Comment (solita-sabinaf) [inline src/helpers/validation.ts:254]: Was cleaning up own console log and thought to take this one along, no idea if it actually was serving any special purpose

File: src/i18n/EnkiIntlProvider.tsx [hunk 5]
```diff
@@ -25,16 +26,27 @@ export const EnkiIntlProvider = ({
...
+          .then((extMessages) => {
+            const combinedMessages = {
+              ...messages,
+              ...extMessages,
+            };
+            setMessages(() => combinedMessages);
+          })
+          .catch();
+      }
     });
```
Comment (solita-sabinaf) [inline src/i18n/EnkiIntlProvider.tsx:40]: Mixing in company-specific translation overrides/additions in here

File: src/i18n/EnkiIntlProvider.tsx [hunk 6]
```diff
@@ -25,16 +26,27 @@ export const EnkiIntlProvider = ({
...
-  }, [selectedLocale]);
+  }, [selectedLocale, partnerCompany]);
 
   return (
     <Suspense>
       {deferredMessages !== null && (
         <IntlProvider
           locale={selectedLocale}
-          messages={deferredMessages}
```
Comment (solita-sabinaf) [inline src/i18n/EnkiIntlProvider.tsx:36]: This felt questionable, there is probably a reason for this that I don't know. But there were problems that when switching the language the translations would get almost always delayed until the next locale change (also sometimes noticeable in the demo env) and with this change it got solved

File: src/index.tsx [hunk 7]
```diff
@@ -48,6 +49,7 @@ const AuthenticatedApp = () => {
   return (
     (authStateLoaded && configStateLoaded && auth.isAuthenticated && (
       <Apollo>
+        <SandboxFeature<{}> feature={`${config.partnerCompany}/CustomStyle`} />
```
Comment (solita-sabinaf) [inline src/index.tsx:52]: Thought this way it would be more future-proof than hardcoding the company name like `feature="FintrafficCustomStyle"`

File: src/scenes/App/NavBar/Logo/index.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        className="logo"
+        src={logo}
+        alt={formatMessage({ id: 'navBarRootLinkLogoAltText' })}
+      />
+      <span>{formatMessage({ id: 'appTitle' })}</span>
+    </div>
+  );
+};
+
+export default Logo;
```
Comment (solita-sabinaf) [inline src/scenes/App/NavBar/Logo/index.tsx:21]: Extracted this into a component so that it can be plugged into the SandboxFeature if the custom logo isn't found

File: src/scenes/App/index.tsx [hunk 9]
```diff
@@ -58,7 +58,23 @@ const App = () => {
...
+              : '/favicon.ico',
+          },
+          {
+            rel: 'manifest',
+            href: partnerCompany
+              ? `src/ext/${partnerCompany}/assets/manifest.json`
+              : '/manifest.json',
+          },
+        ]}
+      />
```
Comment (solita-sabinaf) [inline src/scenes/App/index.tsx:77]: This could also be instead handled at a build time, but felt easier to add this kind of thing

File: src/scenes/Exports/index.tsx [hunk 10]
```diff
@@ -97,11 +97,7 @@ const Exports = () => {
     <div className="exports">
       <Heading1>{formatMessage({ id: 'exportsHeader' })}</Heading1>
 
-      <SecondaryButton
-        as={Link}
-        to="/exports/create"
-        className="create-export-button"
-      >
+      <SecondaryButton as={Link} to="/exports/create" className="create">
         <AddIcon />
```
Comment (solita-sabinaf) [inline src/scenes/Exports/index.tsx:101]: There was a missing space, thought to add it

File: src/config/config.ts [hunk 11]
```diff
@@ -69,4 +71,14 @@ export interface Config {
...
+
+  /**
+   * Name of a company that adopted Nplan.
+   * This is used e.g. for:
+   *    CustomStyle, when determining the relevant custom style class;
+   *    CustomLogo;
+   *    favicon;
+   *    overriding translations (case appTitle)
+   */
+  partnerCompany?: string;
```
Comment (solita-sabinaf) [inline src/config/config.ts:83]: Wasn't sure how to name this, let me know if there is a more fitting term

**Review by testower — CHANGES_REQUESTED** [review]

File: src/config/config.ts (hunk 1 above)
Comment (solita-sabinaf) [reply src/config/config.ts:14]: Yup, definitely! E.g. the the map in the Journey Patterns shall be as a standalone feature, but our custom tiles and default center/zoom level could go under /Fintraffic and then get supplied into the map providers

File: src/i18n/EnkiIntlProvider.tsx (hunk 6 above)
Comment (solita-sabinaf) [reply src/i18n/EnkiIntlProvider.tsx:36]: I've had a look at this again now, and I can't reproduce any problems with deferredMessages when the Fintraffic feature flag is off, it's only an issue when the flag is on. Not clear why, I tried also removing other changes but result was the same. Would you like to have a conditional in there like `messages={partnerCompany ? messages : deferredMessages}`, so the previous way stays same for Entur in case there was some reason for deferredMessages?

File: src/config/config.ts (hunk 11 above)
Comment (solita-sabinaf) [reply src/config/config.ts:83]: Yup, for something company-specific like styles or some translation overrides. Currently looks like: `/ext/Fintraffic/<feature or asset name>`

File: src/ext/SandboxFeature.tsx [hunk 12]
```diff
@@ -43,10 +43,10 @@ const SandboxFeature = <Features, Props extends SandboxFeatureProps<Features>>({
     throw new Error('Max feature depth is 2');
   } else if (splitFeature.length === 2) {
     Component = memo(
-      lazy(() => import(`./${splitFeature[0]}/${splitFeature[1]}/index.ts`)),
+      lazy(() => import(`./${splitFeature[0]}/${splitFeature[1]}/index.tsx`)),
```
[the remark this answers is not in the document]
Comment (solita-sabinaf) [reply src/ext/SandboxFeature.tsx:46]: Should there be an additional prop like "type" or "extension" as part of this, so that this becomes dynamic?
```
export interface SandboxFeatureProps<Features> {
  feature: keyof Features;
  renderFallback?: () => ReactNode;
}
```

[the remark this answers is not in the document]
Comment (testower) [reply src/ext/SandboxFeature.tsx:46]: In some cases you maybe want to export several types from the same file. I think we can just standardize on .tsx since it will always work and it allows to write JSX directly in the index file if you want to do that. My original idea is that the index file would only re-export from other files, hence it was just a .ts.

File: src/i18n/EnkiIntlProvider.tsx (hunk 6 above)
Comment (testower) [reply src/i18n/EnkiIntlProvider.tsx:36]: Like I said, I don't remember why it's there, and if it works without it then we can remove it.

File: src/config/config.ts (hunk 11 above)
Comment (testower) [reply src/config/config.ts:83]: So maybe `whitelabelAssetPath`  or something similar? It highlights what it is used for better I think.

File: src/i18n/EnkiIntlProvider.tsx (hunk 6 above)
Comment (solita-sabinaf) [reply src/i18n/EnkiIntlProvider.tsx:36]: Apparently, after removing the `deferredMessages !== null`, this kind of log shows up 
Screenshot 2024-06-11 at 16 11 43

File: src/config/config.ts (hunk 11 above)
Comment (solita-sabinaf) [reply src/config/config.ts:83]: There was one more kind of use case for partnerCompany, for example in: 
```
 useEffect(() => {
    getMessages(selectedLocale).then(async (messages) => {
      setMessages(() => messages);
      if (partnerCompany) {
        const extMessages = await getExtMessages(partnerCompany, selectedLocale)
        const combinedMessages = {
          ...messages,
          ...extMessages,
        };
        setMessages(() => combinedMessages);
      }
    });
  }, [selectedLocale, partnerCompany]);
```
So it appears also as some indicator that something is meant for non-Entur scenario only.

But this particular case can be reworked into having a SandboxFeature-based custom IntlProvider, while original EnkiIntlProvider then stays entirely intact and partnerCompany then gets renamed into something path related...

Comment (testower) [inline src/config/config.ts:83]: `extPath`

File: package.json [hunk 13]
```diff
@@ -57,7 +57,8 @@
     "react-redux": "8.1.3",
     "react-router-dom": "6.23.1",
     "redux-sentry-middleware": "0.2.2",
-    "uuid": "10.0.0"
+    "uuid": "10.0.0",
+    "@fintraffic/fds-coreui-css": "^0.1.3"
```
Comment (testower) [inline package.json:61]: Please pin this version (remove the ^)

File: src/components/JourneyPatterns/index.tsx [hunk 14]
```diff
@@ -77,7 +77,7 @@ const JourneyPatterns = ({ journeyPatterns, onChange, children }: Props) => {
             })}
             ref={textFieldRef}
           />
-          <div>
+          <div className={'confirm-dialog-buttons'}>
```
Comment (testower) [inline src/components/JourneyPatterns/index.tsx:80]: This change seems unrelated?

File: src/config/config.ts [hunk 15]
```diff
@@ -68,5 +71,14 @@ export interface Config {
   /**
    * Sandbox feature configuration
    */
-  sandboxFeatures?: SandboxFeatures;
+  sandboxFeatures?: SandboxFeatures | TestFeatures;
```
Comment (testower) [inline src/config/config.ts:74]: We should not need to use TestFeatures here

File: src/ext/SandboxFeature.tsx [hunk 16]
```diff
@@ -74,4 +75,7 @@ export default memo(SandboxFeature) as typeof SandboxFeature<
   any
 >;
 
-export const TestSandboxFeature = memo(SandboxFeature) as typeof SandboxFeature;
+export const TestSandboxFeature = memo(SandboxFeature) as typeof SandboxFeature<
```
Comment (testower) [inline src/ext/SandboxFeature.tsx:78]: This type should probably be moved into the test component itself, to avoid any confusion

File: src/scenes/LineEditor/styles.scss [hunk 17]
```diff
@@ -48,3 +48,26 @@
     text-align: right;
   }
 }
+
+.map-container {
```
Comment (testower) [inline src/scenes/LineEditor/styles.scss:52]: Unrelated change?

File: src/scenes/Exports/styles.scss [hunk 18]
```diff
@@ -1,11 +1,13 @@
...
 
 .exports {
   flex: 1;
   display: flex;
   flex-direction: column;
-}
 
-.create-export-button {
-  width: 20rem;
+  .create {
```
Comment (testower) [inline src/scenes/Exports/styles.scss:9]: Unrelated change?

Comment (testower) [reply src/scenes/Exports/styles.scss:9]: Oh I see this is because of https://github.com/entur/enki/pull/1391/files#r1627427420

File: src/config/config.ts (hunk 15 above)
Comment (solita-sabinaf) [reply src/config/config.ts:74]: This was an attempt to resolve this issue during the build:
Screenshot 2024-06-14 at 10 20 00
But maybe should be something else

File: src/index.tsx [hunk 19]
```diff
@@ -66,11 +77,14 @@ const renderIndex = async () => {
     <Sentry.ErrorBoundary>
       <ConfigContext.Provider value={config}>
         <Provider store={store}>
-          <EnkiIntlProvider>
+          <SandboxFeature
```
Comment (testower) [inline src/index.tsx:80]: I see there's an improvement to be made to the SandboxFeature component. the renderFallback should allow to pass props, so that it can take children, that way we don't have to repeat them inside "EnkiApp".

Comment (solita-sabinaf) [reply src/index.tsx:80]: It actually worked with children 😄 I just moved it away to make it look shorter

File: src/scenes/LineEditor/styles.scss (hunk 17 above)
Comment (solita-sabinaf) [reply src/scenes/LineEditor/styles.scss:52]: True that! Gotta move it away

File: src/index.tsx (hunk 19 above)
Comment (solita-sabinaf) [reply src/index.tsx:80]: With the feature flag off, no problems

File: src/config/config.ts (hunk 15 above)
Comment (testower) [reply src/config/config.ts:74]: I think you have these problems because of the changes in src/ext/__tests__/SandboxFeature.test.tsx

This is why I extended the SandboxFeature component type.

I can have a look at the branch if you want

File: src/index.tsx (hunk 19 above)
Comment (solita-sabinaf) [reply src/index.tsx:80]: With feature flag set to false, it does produce this kinds of thing (same as with feature on) when editing a form input on any page
```
@sentry_react.js?v=bca3daac:2942 Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://reactjs.org/link/controlled-components
    at input
    at http://localhost:3001/node_modules/.vite/deps/chunk-57OQC4YP.js?v=fa3f8736:1084:24
    at div
    at InputGroupContextProvider2 (http://localhost:3001/node_modules/.vite/deps/chunk-57OQC4YP.js?v=fa3f8736:649:23)
    at http://localhost:3001/node_modules/.vite/deps/chunk-57OQC4YP.js?v=fa3f8736:691:23
    at http://localhost:3001/node_modules/.vite/deps/chunk-57OQC4YP.js?v=fa3f8736:1051:22
    at section
    at div
    at window.$RefreshReg$ (http://localhost:3001/src/components/GeneralLineEditor/index.tsx:36:3)
    at section
    at LineEditorSteps (http://localhost:3001/src/scenes/LineEditor/LineEditorSteps.tsx:24:11)
    at div
    at div
    at OverlayLoader (http://localhost:3001/src/components/OverlayLoader/index.tsx:22:3)
    at div
    at window.$RefreshReg$ (http://localhost:3001/src/components/LineEditorStepper/index.tsx:30:3)
    at Loading (http://localhost:3001/src/components/Loading/index.tsx:21:3)
    at div
    at div
    at Page (http://localhost:3001/src/components/Page/index.tsx:24:3)
    at commitMutationEffectsOnFiber (http://localhost:3001/node_modules/.vite/deps/chunk-WBIQBQPZ.js?v=fa3f8736:17699:15)
    at RenderedRoute (http://localhost:3001/node_modules/.vite/deps/chunk-ZZJRKN7J.js?v=fa3f8736:3630:5)
    at Routes (http://localhost:3001/node_modules/.vite/deps/chunk-ZZJRKN7J.js?v=fa3f8736:4134:5)
    at div
    at commitMutationEffectsOnFiber (http://localhost:3001/node_modules/.vite/deps/chunk-WBIQBQPZ.js?v=fa3f8736:17699:15)
    at Loading (http://localhost:3001/src/components/Loading/index.tsx:21:3)
    at div
    at div
    at div
    at ScrollToTop (http://localhost:3001/src/components/ScrollToTop/index.tsx:20:24)
    at Router (http://localhost:3001/node_modules/.vite/deps/chunk-ZZJRKN7J.js?v=fa3f8736:4077:15)
    at BrowserRouter (http://localhost:3001/node_modules/.vite/deps/react-router-dom.js?v=0683f181:549:5)
    at div
    at commitMutationEffectsOnFiber (http://localhost:3001/node_modules/.vite/deps/chunk-WBIQBQPZ.js?v=fa3f8736:17699:15)
    at ApolloProvider (http://localhost:3001/node_modules/.vite/deps/@apollo_client.js?v=a8152831:6238:19)
    at Apollo (http://localhost:3001/src/api/index.tsx:69:26)
    at commitMutationEffectsOnFiber (http://localhost:3001/node_modules/.vite/deps/chunk-WBIQBQPZ.js?v=fa3f8736:17699:15)
    at AuthProvider (http://localhost:3001/node_modules/.vite/deps/react-oidc-context.js?v=572331a2:3188:5)
    at AuthProvider (http://localhost:3001/src/auth/auth.tsx:106:32)
    at commitMutationEffectsOnFiber (http://localhost:3001/node_modules/.vite/deps/chunk-WBIQBQPZ.js?v=fa3f8736:17699:15)
    at Suspense
    at EnkiIntlProvider (http://localhost:3001/src/i18n/EnkiIntlProvider.tsx:25:3)
    at EnkiApp
    at Suspense
    at SandboxFeature (http://localhost:3001/src/ext/SandboxFeature.tsx:21:3)
    at Provider (http://localhost:3001/node_modules/.vite/deps/react-redux.js?v=a7105bce:1253:3)
    at commitMutationEffectsOnFiber (http://localhost:3001/node_modules/.vite/deps/chunk-WBIQBQPZ.js?v=fa3f8736:17699:15)
```

Comment (testower) [reply src/index.tsx:80]: I think this warning is unrelated

Comment (solita-sabinaf) [reply src/index.tsx:80]: But the stuff above is actually reproducing in the master branch as well, so at least not related to these changes

File: src/config/config.ts (hunk 15 above)
Comment (solita-sabinaf) [reply src/config/config.ts:74]: But it was also occurring earlier, e.g. at the time of commit 2823a87145f514e9e28826ed9b7f11d35ea1bc9b, in a way that's how that chain of changes in the test started

File: src/components/JourneyPatterns/index.tsx (hunk 14 above)
Comment (solita-sabinaf) [reply src/components/JourneyPatterns/index.tsx:80]: This was helping with buttons in this dialog, just came along the way while making custom styles
Screenshot 2024-06-14 at 14 16 36
Screenshot 2024-06-14 at 14 16 44

Comment (testower) [reply src/components/JourneyPatterns/index.tsx:80]: Ah nice, thanks for clarifying

Comment (solita-sabinaf) [reply src/components/JourneyPatterns/index.tsx:80]: This is actually not an issue in the Entur styling, but it hepled in our case and it doesn't cause problems in the normal scenario (that class is used in other dialogs too)

File: src/config/config.ts (hunk 15 above)
Comment (testower) [reply src/config/config.ts:74]: Here's a commit that isolates test related code in test files https://github.com/entur/enki/commit/af4be048ec8cab58a16b83fc8a6eaa3786694715

It's acceptable to ignore the typescript error on the config provider, since it's not the part that's being tested here.

File: src/config/config.ts [hunk 20]
```diff
@@ -9,7 +10,9 @@ import { Locale } from '../i18n';
  *  For multi-level features, only the top-level featureName should be
  *  toggled.
  */
-export interface SandboxFeatures {}
+export interface SandboxFeatures {
+  Fintraffic: boolean;
```
Comment (testower) [inline src/config/config.ts:14]: It would be nice to document each sandbox feature with a comment explaining what it is

**Comment by testower** [comment]
I think we're on the home stretch now, just some minor things left 🎉

File: src/config/config.ts (hunk 15 above)
Comment (solita-sabinaf) [reply src/config/config.ts:74]: Looks great 👍 Got it cherry-picked

**Review by testower — APPROVED** [review]
